### PR TITLE
Do not commit after ARG dockerfile command

### PR DESCRIPTION
**Pull request by duglin**
Since the ARG Dockerfile command doesn't change anything in the image
we're building there's no point in committing anything after processing
the command.

Signed-off-by: Doug Davis <dug@us.ibm.com>

### CODE REVIEW

**Review by vdemeester — COMMENTED** [review]
SGTM
/cc @thaJeztah @tiborvass

File: integration-cli/docker_cli_build_test.go [hunk 1]
```diff
@@ -5694,8 +5694,39 @@ func (s *DockerSuite) TestBuildBuildTimeArgHistory(c *check.C) {
...
+	// Should not see foo in history
+	if strings.Contains(outputTabs, envKey) {
+		c.Fatalf("found arg in image history output: %q", outputTabs)
+	}
+}
+
+func (s *DockerSuite) TestBuildBuildTimeArgIDs(c *check.C) {
+	imgName := "bldargtest"
+
+	dockerfile1 := fmt.Sprintf("FROM busybox")
```
Comment (AkihiroSuda) [inline integration-cli/docker_cli_build_test.go]: nit: no need to use Sprintf here

Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: ha! :-) that what happens with blind copy-n-pastes.  Fixed. thanks

**Comment by clnperez** [comment]
@duglin, if you're using the intermediary containers to debug a build, would these still be set? i.e.

```
FROM foo
ENV bar true
RUN failing command
```

Would `bar` be set to `true` in a container that I could fire up and play around with `failing command`?

**Comment by clnperez** [comment]
@duglin My brain misread ARG as ENV. So sorry. (And yes, with that adjustment, it makes perfect sense). 😆

**Comment by tonistiigi** [comment]
Would it make sense to keep it just so that it shows up in the history? cc @tianon

**Comment by duglin** [comment]
@tonistiigi that's an interesting question.  So take this example:
```
FROM ubuntu
ARG foo=bar
RUN env
```
With a build command of: `docker build --build-arg foo=zoo .`
The history will show:
```
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
ba51e206c3d2        4 seconds ago       |1 foo=zoo /bin/sh -c env                       0 B                 
12a549573c9a        29 minutes ago      /bin/sh -c #(nop)  ARG foo=bar                  0 B                 
...
```
The ARG line in the history might actually be kind of misleading because "bar" is not the value of "foo".  Granted, I don't think we encourage people to look at the `|1 foo=zoo` part of the history for info, but I think that's more accurate than what they'll see in the ARG line.

**Comment by LK4D4** [comment]
@duglin I'm not quite sure if you agree with @tonistiigi or not :) So I'm pinging both of you.

**Comment by duglin** [comment]
@LK4D4 sorry I wasn't clear :-)   I'm not agreeing with him. I don't think there's value in showing ARG in the image history in fact it could be misleading (ie. show misleading/wrong info) as my example shows.

**Comment by tonistiigi** [comment]
@tianon You did some work for restoring Dockerfile based on history that this should break. If you don't see an issue with that then this change is fine by me.

**Comment by tianon** [comment]
Thanks for the prod @tonistiigi :heart:

IMO this should really go the other way -- commit, but commit with the value the argument ended up having (rather than the value from the `Dockerfile`)

In the end, it's probably harmless either way, since `ARG`s make the following instructions really hard to parse (given that the `|` hacks do _not_ stringify well), but it seems odd to remove this from history and not other things like `ENV` that are functionally similar, especially since the explicit empty commit doesn't cost nearly as much as it used to (especially in transport, where it's essentially optimized completely away now).

**Comment by thaJeztah** [comment]
I don't feel comfortable changing this, as it would only bring a minor improvement in performance, but we kinda "break" the current assumption that each instruction in the Dockerfile results in a history entry.

I agree that the history item is fairly useless, but perhaps there's a tool out there using the history entries that would break by this.
